silq/tests/test_pulse_sequences.py

```diff
@@ -803,5 +1023,505 @@ def test_pickle_pulse_sequence_two_pulses(self):
+             ]
+        ]
+        for N_pulses in range(4):
+            for pulses in itertools.product(possible_pulses, repeat=N_pulses):
+                pulse_sequence.settings.RF_pulses = pulses
+                pulse_sequence.generate()
+
+                stage_pulses = pulse_sequence.get_pulses('plunge')
+                read_pulses = pulse_sequence.get_pulses('read')
+                ESR_pulses = iter(pulse_sequence.get_pulses('ESR'))
+
+                t = 0
+                for k in range(len(pulses)):
+                    stage_pulse = stage_pulses[k]
+                    read_pulse = read_pulses[k]
+                    RF_pulses = pulses[k]
+                    if RF_pulses is None:
+                        RF_pulses = []
+                    elif isinstance(RF_pulses, Pulse):
+                        RF_pulses = [RF_pulses]
+
+                    self.assertAlmostEqual(stage_pulse.t_start, t)
+
+                    t += pulse_sequence.settings.pre_delay
+
+                    for k, RF_pulse in enumerate(RF_pulses):
+                        ESR_pulse = next(ESR_pulses)
+                        self.assertAlmostEqual(ESR_pulse.duration, RF_pulse.duration)
+                        self.assertAlmostEqual(ESR_pulse.t_start, t)
+
+                        t += ESR_pulse.duration
+                        if k < len(RF_pulses) - 1:
+                            t += pulse_sequence.settings.inter_delay
+
+                    t += pulse_sequence.settings.post_delay
+
+                    self.assertAlmostEqual(stage_pulse.t_stop, t)
+                    self.assertAlmostEqual(read_pulse.t_start, t)
+
+                    self.assertAlmostEqual(
+                        read_pulse.duration,
+                        pulse_sequence.settings.read_pulse.duration
+                    )
+                    t += read_pulse.duration
+
+    def test_RF_pulseSequence(self):
+        pulse_sequence = self.pulse_sequence
+        RF_pulse_sequence = PulseSequence(
+            name='RF_subsequence',
+            pulses=[
+                SinePulse('sub_ESR', t_start=0.3e-3, duration=2e-3)
+            ]
+        )
+        RF_pulse_sequence.duration = 5e-3
+
+        pulse_sequence.settings.RF_pulses = [
+            SinePulse('ESR', duration=1.5e-3),
+            RF_pulse_sequence
+        ]
+
+        pulse_sequence.generate()
+
+        stage_pulses = pulse_sequence.get_pulses('plunge')
+        ESR_pulse = pulse_sequence.get_pulse('ESR')
+        sub_ESR_pulse = pulse_sequence.get_pulse('sub_ESR')
+
+        self.assertEqual(stage_pulses[0].t_start, 0)
+        self.assertEqual(ESR_pulse.t_start, 0.1e-3)
+        self.assertEqual(ESR_pulse.t_stop, 1.6e-3)
+        self.assertEqual(stage_pulses[0].t_stop, 1.9e-3)
+
+        self.assertEqual(stage_pulses[1].t_start, 2.9e-3)
+        self.assertEqual(sub_ESR_pulse.t_start, 3.2e-3)
+        self.assertEqual(sub_ESR_pulse.t_stop, 5.2e-3)
+        self.assertEqual(stage_pulses[1].t_stop, 7.9e-3)
+
+
```

Comment:
what are these tests? it's like your testing ground? Do you want to merge into master/main branch?

Reply:
No these are unit tests. They are a programming technique to help identify bugs and check if new changes break anything. See https://realpython.com/python-testing/ for more information

Reply:
ok